modules/library/referencing/src/main/java/org/geotools/referencing/operation/matrix/GeneralMatrix.java

```diff
@@ -569,28 +630,22 @@ static boolean isIdentity(final Matrix matrix, double tolerance) {
      * {@inheritDoc}
      */
     public final void multiply(final Matrix matrix) {
-        final GeneralMatrix m;
-        if (matrix instanceof GeneralMatrix) {
-            m = (GeneralMatrix) matrix;
-        } else {
-            m = new GeneralMatrix(matrix);
-        }
-        mul(m);
+        mul(matrix);
     }
 
-    
-
     @Override
     public int hashCode() {
         final int prime = 31;
         int result = 1;
-        
-        if( mat == null ){
+        if (mat == null) {
             return prime * result;
         }
         result = prime * result + mat.numRows;
         result = prime * result + mat.numCols;
-        result = prime * result + (int) mat.data[0];
+        for (double d : mat.data) {
+            long bits = Double.doubleToRawLongBits(d);
+            result = prime * ((int)(bits ^ (bits >>> 32)));
+        }
```

Comment:
> Confused about the new hashcode business in GeneralMatrix:
> Issues:
> * This does not seem like a valid duplicate of the original hash code of vecmath, that used all matrix elements (this one will result in a accumulation of all matrices having the same top/left element in the same bucket)
> * The code is duplicated in a static method a few lines below:
> This static method is then called to build the hashcode of Matrix3, but I don't see the reason, aren't we inheriting the GeneralMatrix one anyways? Same goes for equals with tolerance in Matrix3, seems that we should be inheriting the base class one

Hashcode implementation was generated by hand without reference to vecmath. I have had a second attempt at implementing now that I know we would like to reference every element.

The remaining duplication was introduced by having Matrix3 extend GeneralMatrix and has now been resolved.